Casting the sharing_group_id to int value

I-am-Sherlocked opened the pull request:
Saving the sharing_group_id as it is returned by CaptureSG results in 

> Error: [PDOException] SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'sharing_group_id' at row 1. 

Wrapping it in intval will insert the correct int value.

I-am-Sherlocked committed:
Casting the sharing_group_id to int value 

Saving the sharing_group_id as it is returned by CaptureSG results in Error: [PDOException] SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'sharing_group_id' at row 1. Wrapping it in intval will insert the correct int value.

I-am-Sherlocked commented:
This is happening when CaptureSG in SharingGroup.pgp returns False, the value is converted to an empty string by php.

iglocska commented:
Interesting, need to look into this. If it returns false, we should block the event save process.

sfossen commented:
Looking at it, there are 3 cases:
1. unable to save the sharing group ( SharingGroup->save() )
2. sharing group without enough information to create one ( missing ['Organisation'] and ['SharingGroupOrg'] )
3. if the user is not authorized to see the group (SharingGroup->checkIfAuthorised() )

I'm experiencing 2 in a MISP import of a MISP export in XML, this fix allows the import to complete In the first branch of the SharingGroup->captureSG() it throws an exception when the user is not authorized to create a SharingGroup, perhaps  3 should as well.
And distribution should change to 0 for 2 ?

I-am-Sherlocked committed:
Fixing the error caused by a false sharing_group_id

If SharingGroup->captureSG returned false indicating it did not save the sharing group, then distribution should be set to 0, and the sharing_group_id to an integer 0.

I-am-Sherlocked committed:
Throw exception for sharing group if unauthorised user

Instead of returning a false value for sharing_group_id, throw an exception if user is not authorised to save modifications to that sharing group.

iglocska commented:
2 is tricky. Converting it to 0 gives it a different meaning, so I'm a bit torn whether we should do that or outright block it and throw an exception.

sfossen commented:
Looking at the source MISP where the data is coming from.

There is an internal MISP sharing to an external MISP through an internal sharing group. The external MISP created a passive sharing group, and both the export and sync operations create sharing group information that is not importable nor syncable.

The newly proposed fixes, allow the both operations to proceed to completion. Looking at the permissions of the passive sharing group, it is supposed to be accessible and syncable to all orgs on that MISP.

sfossen commented:
@iglocska please comment again, a fix is needed or different proposal.

iglocska commented:
Sorry will try to look into this later tonight.

iglocska commented:
Can't see anything wrong with it, if issues pop-up we can look at it again. Thanks a lot for the PR and sorry for the slow response.
